sim/firesim-lib/src/main/cc/bridges/tracerv.cc

```diff
@@ -162,93 +176,97 @@ void tracerv_t::init() {
-                        fprintf(this->tracefiles[0], "%016lx", OUTBUF[i+5]);
-                        fprintf(this->tracefiles[0], "%016lx", OUTBUF[i+4]);
-                        fprintf(this->tracefiles[0], "%016lx", OUTBUF[i+3]);
-                        fprintf(this->tracefiles[0], "%016lx", OUTBUF[i+2]);
-                        fprintf(this->tracefiles[0], "%016lx", OUTBUF[i+1]);
-                        fprintf(this->tracefiles[0], "%016lx\n", OUTBUF[i+0]);
-                    } else {
-                        for (int q = 0; q < NUM_CORES; q++) {
-                           if ((OUTBUF[i+0+q] >> 40) & 0x1) {
-                             fprintf(this->tracefiles[q], "C%d: %016llx, cycle: %016llx\n", q, OUTBUF[i+0+q], OUTBUF[i+7]);
-                           }
-                        }
+    pull(dma_addr, (char*)OUTBUF, num_beats * 64);
+    //check that a tracefile exists (one is enough) since the manager
+    //does not create a tracefile when trace_enable is disabled, but the
+    //TracerV bridge still exists, and no tracefile is created by default.
+    if (this->tracefile) {
+        if (this->human_readable || this->test_output) {
+            for (int i = 0; i < num_beats * 8; i+=8) {
+                if (this->test_output) {
+                    fprintf(this->tracefile, "%016lx", OUTBUF[i+7]);
+                    fprintf(this->tracefile, "%016lx", OUTBUF[i+6]);
+                    fprintf(this->tracefile, "%016lx", OUTBUF[i+5]);
+                    fprintf(this->tracefile, "%016lx", OUTBUF[i+4]);
+                    fprintf(this->tracefile, "%016lx", OUTBUF[i+3]);
+                    fprintf(this->tracefile, "%016lx", OUTBUF[i+2]);
+                    fprintf(this->tracefile, "%016lx", OUTBUF[i+1]);
+                    fprintf(this->tracefile, "%016lx\n", OUTBUF[i+0]);
+                // At least one valid instruction
+                } else {
+                    for (int q = 0; q < core_ipc; q++) {
+                       if (OUTBUF[i+q+1] & valid_mask) {
+                           fprintf(this->tracefile, "Cycle: %016lld I%d: %016llx\n", OUTBUF[i], q, OUTBUF[i+q+1] & (~valid_mask));
+                       } else {
+                           break;
+                       }
                     }
                 }
-            } else if (this->fireperf) {
-                for (int i = 0; i < QUEUE_DEPTH * 8; i+=8) {
-                    uint64_t cycle_internal = OUTBUF[i+7];
+            }
+        } else if (this->fireperf) {
+
+            for (int i = 0; i < QUEUE_DEPTH * 8; i+=8) {
+                uint64_t cycle_internal = OUTBUF[i+7];
 
-                    for (int q = 0; q < NUM_CORES; q++) {
-                        if ((OUTBUF[i+0+q] >> 40) & 0x1) {
-                            uint64_t iaddr = (uint64_t)((((int64_t)(OUTBUF[i+0+q])) << 24) >> 24);
-                            this->trace_trackers[q]->addInstruction(iaddr, cycle_internal);
+                for (int q = 0; q < core_ipc; q++) {
```

Comment:
did we check the order of insts here vs. program order? 

unless we've tested this, it might be a good idea to stick an assert in here to require core_ipc to be 1 for fireperf mode until someone tests it. just so it's not silently broken.

CC: @abejgonzalez @jerryz123

Reply:
I looked at a couple traces and they seemed fine, but that's hardly proof.